ssl/statem/statem_clnt.c

```diff
@@ -3814,19 +3828,40 @@ CON_FUNC_RETURN tls_construct_client_certificate(SSL_CONNECTION *s,
      * If we attempted to write early data or we're in middlebox compat mode
      * then we deferred changing the handshake write keys to the last possible
      * moment. We need to do it now.
+     * NOTE: We split this into quic and non-quic cases here to suppress the
+     * cipher state change if SSL_EARLY_DATA_NONE is set in the early data state
+     * to create a single update of the keys in conjunction with tls_process_server_hello
      */
-    if (SSL_CONNECTION_IS_TLS13(s)
+    if (SSL_IS_QUIC_HANDSHAKE(s)) {
+        if (SSL_CONNECTION_IS_TLS13(s)
             && SSL_IS_FIRST_HANDSHAKE(s)
             && (s->early_data_state != SSL_EARLY_DATA_NONE
                 || (s->options & SSL_OP_ENABLE_MIDDLEBOX_COMPAT) != 0)
+        && s->write_key_yielded == 0
```

Comment:
Indentation is off here. But will this ever be true? Won't we always have yielded the write key in the process_server_hello prior to this?

Reply:
Yeah, that seems true, we will never send a client certificate before we process a server hello, I'll remove this